### PR TITLE
Prestissimo: Fix logic deserializing LongDecimal-typed Data Encoded in INT128_ARRAY in bytestream read

### PR DESCRIPTION
**Pull request by markjin1990**
== RELEASE NOTES ==

General Changes
* On the Velox side, reading int128 in Bytestream is not yet supported, and instead, it reads two 64bit integers and combine them for form an int128. However, on Prestissimo, the loophole seems to still exist. Without this fix, in the release mode, Prestissimo will crash with EXC_BAD_ACCESS error when deserializing LongDecimal data (unit tests) since it is trying to directly parse a 128-bit block.

This is the backtrace information of the core dump we encounter.
image

Test plan
* Added multiple unit tests for deserializing base64 encoded long Decimals for Prestissimo

### CODE REVIEW

**Comment by markjin1990** [comment]
@majetideepak Appreciate if you could help review this PR when you have bandwidth.

**Comment by yzhang1991** [comment]
Hi @aditi-pandit @karteekmurthys can you help take a look at this since Deepak is still on parental leave?

**Comment by karteekmurthys** [comment]
@markjin1990 Thanks for taking this up. I am not sure why we didn't face this issue. We have e2e Decimal query tests that use 128 bit numbers: https://github.com/prestodb/presto/pull/18320/files#diff-efd1daf1b9c66dbc28260d458ec2ddc109f06cdbf9db2aa0696f70c785636401R452

We are able to successfully parse incoming 128-bit stream from Presto co-ordinator to Velox Expressions in PrestoCPP. We don't parse 128-bit data stream as is because the 128-bit representation uses signed magnitude representation. Please refer this issue: https://github.com/prestodb/presto/issues/18312

Just to make sure, I copied the unit test from this PR to existing code base and the tests pass without your changes. So, not sure if your fix has any impact. Lmk if I am missing something.

Screen Shot 2022-12-26 at 9 31 40 PM

**Comment by markjin1990** [comment]
> Just to make sure, I copied the unit test from this PR to existing code base and the tests pass without your changes. So, not sure if your fix has any impact. Lmk if I am missing something.

@karteekmurthys Thanks for your reply! Did you test in debug mode or release mode? Debug mode should be fine, but release mode gives me the "sigsegv - [EXC_BAD_ACCESS (code=EXC_I386_GPFLT)]" error. Clion crashed at this line in disassembly.
(https://stackoverflow.com/questions/19651788/whats-the-meaning-of-exception-code-exc-i386-gpflt)
image

**Comment by karteekmurthys** [comment]
> Debug mode should be fine, but release mode gives me the "sigsegv - [EXC_BAD_ACCESS (code=EXC_I386_GPFLT)]" error. Clion crashed at this line in disassembly.

I had run both, but shared the debug mode results. Here are the release mode results as well:
Screen Shot 2022-12-26 at 9 49 53 PM

**Comment by markjin1990** [comment]
> > Debug mode should be fine, but release mode gives me the "sigsegv - [EXC_BAD_ACCESS (code=EXC_I386_GPFLT)]" error. Clion crashed at this line in disassembly.
> 
> I had run both, but shared the debug mode results. Here are the release mode results as well:
> 
> Screen Shot 2022-12-26 at 9 54 05 PM
> 
> Edit: Updated the snapshot.

Hmm, weird. I just tested on the latest presto repo in my local machine and the cloud environment and the seg fault issue still exists
image

This is my local machine info output by `velox/script/info.sh`.

```
Velox System Info v0.0.2
Commit: 4a0b62e4aa6051a79b2247c50c3eb15cfeb2baef
CMake Version: 3.25.1
System: Darwin-21.6.0
Arch: x86_64
C++ Compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
C++ Compiler Version: 13.0.0.13000029
C Compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
C Compiler Version: 13.0.0.13000029
CMake Prefix Path: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr;/usr/local;/usr;/;/usr/local/Cellar/cmake/3.25.1;/usr/local;/usr/X11R6;/usr/pkg;/opt;/sw;/opt/local
```

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp [hunk 1]
```diff
@@ -82,6 +82,47 @@ TEST_F(Base64Test, singleTinyint) {
...
+      "DAAAAElOVDEyOF9BUlJBWQEAAAAAZAAAAAAAAAAAAAAAAAAAgA==";
+  auto vector2 = readBlock(LONG_DECIMAL(24, 2), data2, pool_.get());
+
+  ASSERT_EQ(TypeKind::LONG_DECIMAL, vector2->typeKind());
+  ASSERT_EQ(1, vector2->size());
+  ASSERT_FALSE(vector2->isNullAt(0));
+
+  auto decimalVector2 =
+      vector2->as<FlatVector<facebook::velox::UnscaledLongDecimal>>();
+  ASSERT_EQ(UnscaledLongDecimal(-100), decimalVector2->valueAt(0));
```
Comment (karteekmurthys) [inline presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:123]: Add bigger values (> 64 bit) to the tests with positive and negative values.

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp [hunk 2]
```diff
@@ -82,6 +82,47 @@ TEST_F(Base64Test, singleTinyint) {
   ASSERT_EQ(1, intVector->valueAt(0));
 }
 
+TEST_F(Base64Test, simpleLongDecimal) {
+  // Unscaled value = 0
+  const std::string data0 =
+      "DAAAAElOVDEyOF9BUlJBWQEAAAAAAAAAAAAAAAAAAAAAAAAAAA==";
```
Comment (majetideepak) [inline presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:88]: These string values seem cryptic. Can we directly cast the unscaled values?

**Review by majetideepak — COMMENTED** [review]
int128_t pointer casts can cause segfaults on unaligned memory blocks. This change looks good to me.

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp (hunk 2 above)
Comment (majetideepak) [reply presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:88]: I guess there is metadata involved. Can we construct the payload explicitly?

**Comment by karteekmurthys** [comment]
@markjin1990 would you please address the review comments and help close this PR?

**Comment by markjin1990** [comment]
@karteekmurthys Absolutely! Will do tonight.

**Comment by markjin1990** [comment]
Hi, @karteekmurthys! I added new unit tests of large values (>64bit) per your suggestion. Can we merge it?

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp (hunk 2 above)
Comment (karteekmurthys) [reply presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:88]: @markjin1990 Did you push changes to address this comment? It would be easier to verify what exact value been passed in the encoded string if you construct the payload.

**Review by karteekmurthys — APPROVED** [review]
LGTM!

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp (hunk 2 above)
Comment (majetideepak) [reply presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:88]: @markjin1990 how are these payloads even built now? Can you add a comment?

**Comment by majetideepak** [comment]
@markjin1990 can you also rebase with the master branch?

**Comment by markjin1990** [comment]
> @markjin1990 can you also rebase with the master branch?

Just did. Thanks for noticing

File: presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp (hunk 2 above)
Comment (markjin1990) [reply presto-native-execution/presto_cpp/presto_protocol/tests/Base64Test.cpp:92]: > @markjin1990 how are these payloads even built now? Can you add a comment?

Just added. I basically synthesized some records of the corresponding decimal values in the table and print the value of the 2nd parameter of `base64Encoded` of `readBlock` to the screen in online running service of Prestissimo when reading these synthesized values. Hope that is clear. 

Sorry that I haven't found an easy approach automatically synthesizing these values within the unit test and nor the other unit tests have demonstrated their approaches getting these values in the same test file.

**Comment by majetideepak** [comment]
@markjin1990 It still says this branch is out-of-date. Can you pull the latest master and rebase?
Also, please squash all the commits into a single commit. Please prefix [native] to the commit and title as described here https://github.com/prestodb/presto/tree/master/presto-native-execution#creating-prs-for-prestopresto-native-execution.

**Comment by markjin1990** [comment]
> @markjin1990 It still says this branch is out-of-date. Can you pull the latest master and rebase? Also, please squash all the commits into a single commit. Please prefix [native] to the commit and title as described here https://github.com/prestodb/presto/tree/master/presto-native-execution#creating-prs-for-prestopresto-native-execution.

@majetideepak Thanks for the heads up! It's resolved.

**Comment by majetideepak** [comment]
@markjin1990 it still says out-of-date to me. Do you see the same?
And can you fix the commit message? Probably make the commit title
```
[native] Fix deserializing of encoded INT128_ARRAY in byte-stream read
```
 
and make the commit message 
```
Prestissimo crashes with EXC_BAD_ACCESS error when deserializing LongDecimal data
as it tries to parse a 128-bit value. This is now fixed by constructing the 128-bit value after
reading two 64-bit values.
```

**Comment by markjin1990** [comment]
> @markjin1990 it still says out-of-date to me. Do you see the same? And can you fix the commit message? Probably make the commit title
> 
> ```
> [native] Fix deserializing of encoded INT128_ARRAY in byte-stream read
> ```
> 
> and make the commit message
> 
> ```
> Prestissimo crashes with EXC_BAD_ACCESS error when deserializing LongDecimal data
> as it tries to parse a 128-bit value. This is now fixed by constructing the 128-bit value after
> reading two 64-bit values.
> ```

@majetideepak fixed the commit msg and rebased again, thanks!

**Review by majetideepak — APPROVED** [review]

**Comment by markjin1990** [comment]
Hi! @mbasmanova Would you kindly help merging this PR to the master branch? @majetideepak and @karteekmurthys already reviewed it. Thanks!

**Comment by mbasmanova** [comment]
Should we add an e2e test that would crash/fail without the fix?

**Comment by majetideepak** [comment]
@mbasmanova: @markjin1990 is seeing this with existing tests on MacOS.

```
Velox System Info v0.0.2
Commit: 4a0b62e4aa6051a79b2247c50c3eb15cfeb2baef
CMake Version: 3.25.1
System: Darwin-21.6.0
Arch: x86_64
C++ Compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
C++ Compiler Version: 13.0.0.13000029
C Compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
C Compiler Version: 13.0.0.13000029
CMake Prefix Path: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr;/usr/local;/usr;/;/usr/local/Cellar/cmake/3.25.1;/usr/local;/usr/X11R6;/usr/pkg;/opt;/sw;/opt/local
```

**Comment by majetideepak** [comment]
See comment here https://github.com/prestodb/presto/pull/18838#issuecomment-1366785521

**Comment by mbasmanova** [comment]
@majetideepak Got it. Thank you for clarifying,

@markjin1990 I'm seeing "This branch is out-of-date with the base branch" message. Please, rebase and ping when tests are green.

**Comment by majetideepak** [comment]
@markjin1990 It still says the branch is out-of-date

**Comment by markjin1990** [comment]
> @markjin1990 I'm seeing "This branch is out-of-date with the base branch" message. Please, rebase and ping when tests are green.

@mbasmanova rebased and tests are now complete except for "continuous-integration/jenkins/pr-merge", which I think has nothing to do with my PR.

**Comment by mbasmanova** [comment]
@markjin1990 Thanks.
